[smaenergymeter] Fix handling of broadcast frames

splatch opened the pull request:
# SMA Energy Meter improvements

There are two outstanding issues which are closely related to how UDP traffic is handled by SMA energy meter binding.
This PR addresses both by separating several concerns into their own places, separating network communication from data parsing and port management. Starting point for fix were odd values reported by meter binding in #11497. Additionally it introduces support for handling of multiple meters requested in #3429 which is impossible prior provided fixes.

This work wouldn't be possible without amazing analysis made by Thomas: https://community.openhab.org/t/sma-energy-meter-binding-yields-unplausible-values/128180/4.

monnimeter reviewed (approved):
👍 Thank you much!

splatch commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java:
```diff
@@ -0,0 +1,130 @@
+                        handler.handle(meter);
+                    }
+                } catch (IOException e) {
+                    logger.info("Unexpected payload received for group {}", group, e);
```
Should be lowered to debug or trace level!

kaikreuzer replied:
Indeed.

kaikreuzer commented on bundles/org.openhab.binding.smaenergymeter/src/main/resources/OH-INF/thing/energyMeter.xml:
```diff
@@ -32,6 +32,10 @@
 		<config-description>
+			<parameter name="serialNumber" type="text" required="true">
+				<label>Serial number</label>
+				<description>Identifier of meter </description>
```
```suggestion
				<description>Identifier of meter</description>
```

splatch commented:
I've updated PR as there are some troubles with amount of changes I introduced. We're clearing issues out in forum topic mentioned earlier.

lolodomo commented:
@splatch : do you plan to finish that PR or should we close it ? In case you expect a review, please fix the conflicts and make the PR valid (it includes unexpected changes in many bindings).

splatch commented:
@lolodomo I've updated branch, however not tested it yet as I have no 3.4.0 deployment with sma available. Lets see if others can confirm it working.

lolodomo commented:
What about the flag "work in progress" ? I understand you did not test your PR, are you enough confident with your changes to remove the "WIP" flag ?
@monnimeter: could you please have a look and confirm that the current PR code is still OK for you ?

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/EnergyMeter.java:
```diff
@@ -81,23 +71,21 @@ public EnergyMeter(String multicastGroup, int port) {
+                        "Received frame with wrong protocol ID " + Arrays.toString(protocolId));
+            }
 
+            byte[] idf = Arrays.copyOfRange(bytes, 0x14, 0x18);
```
`idf` is apparently not used.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/EnergyMeter.java:
```diff
@@ -81,23 +71,21 @@ public EnergyMeter(String multicastGroup, int port) {
-            String sma = new String(Arrays.copyOfRange(bytes, 0x00, 0x03));
+    public void parse(byte[] bytes) throws IOException {
+        try {
+            String sma = new String(Arrays.copyOfRange(bytes, 0, 3));
```
Maybe you could start by checking that bytes contains enough data ?
At least 0x18 ?

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +74,121 @@ public void initialize() {
+                if (!thing.getProperties().containsKey(Thing.PROPERTY_SERIAL_NUMBER)) {
+                    Map<String, String> props = editProperties();
+                    props.put(Thing.PROPERTY_VENDOR, "SMA");
+                    props.put(Thing.PROPERTY_SERIAL_NUMBER, serialNumber);
```
You can't have at the same time a configuration parameter and a property with the same name.
As you add in that PR a configuration parameter, do not set anymore a property.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +74,121 @@ public void initialize() {
+            PacketListener thingListener = listenerRegistry.getListener(config.getMcastGroup(), port);
+            serialNumber = config.getSerialNumber();
+            if (serialNumber == null) {
+                if (!thing.getProperties().containsKey(Thing.PROPERTY_SERIAL_NUMBER)
+                        || (serialNumber = retrieveSerialNumber(thing.getProperties())) == null) {
```
No need to check in properties, serialNumber  will be set by the new configuration parameter you added to the thing.

Normally, the recent core framework should check that `config.getSerialNumber()` is not null for you as you defined the parameter as required. But keep you can keep the check.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +74,121 @@ public void initialize() {
+                    return;
+                }
+
+                // copy serial number from thing properties into config
+                Map<String, Object> newConfig = editConfiguration().getProperties();
+                newConfig.put("serialNumber", serialNumber);
+                Thing thing = editThing().withConfiguration(new Configuration(newConfig)).build();
+                updateThing(thing);
```
All this block can be removed.

Do you try to handle with old things ?

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/resources/OH-INF/thing/energyMeter.xml:
```diff
@@ -32,6 +32,10 @@
 		</properties>
 
 		<config-description>
+			<parameter name="serialNumber" type="text" required="true">
```
You will have to update the file smaenergymeter.properties.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +74,121 @@ public void initialize() {
+                    TimeUnit.MILLISECONDS);
+            this.listener = thingListener;
+            thingListener.addPayloadHandler(this);
+            thingListener.open();
```
The code run by `initialize` should be fast. Are you sure that this "open" will return fast ? If not, this is better to run code that can take time in a separate thread.

splatch replied:
Its udp code so there is only local network stack and no remote interactions we wait for.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +74,121 @@ public void initialize() {
+            thingListener.open();
+            logger.debug("Listening to meter updates and publishing its measurements every {}ms for '{}'",
+                    updateInterval, getThing().getUID());
+            // we do not set online status here, it will be set only when data is received
```
initialize should always set a thing status before returning.
Are you certain that you will receive a data and immediately after ? If not, set the thing status to UNKNOWN and then to ONLINE when a data will be received.

splatch replied:
Few lines above there is a scheduled task which will mark thing offline if packet does not arrive within two poll cycles.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +74,121 @@ public void initialize() {
+
+            this.updateInterval = TimeUnit.SECONDS
+                    .toMillis(config.getPollingPeriod() == null ? 30 : config.getPollingPeriod());
+            timeoutFuture = scheduler.scheduleWithFixedDelay(this, updateInterval * 2, updateInterval,
+                    TimeUnit.MILLISECONDS);
```
Rather than calling this with you class extending Runnable, just call your method `run` there and no need to implement Runnable.

splatch replied:
For clarity of stack trace I prefer implementing interfaces and not having lambdas or anonymous clases with generated names.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +74,121 @@ public void initialize() {
+        long currentTimeMillis = System.currentTimeMillis();
+        if (currentTimeMillis - lastUpdate > updateInterval * 2) {
+            long missedWindow = TimeUnit.MILLISECONDS.toSeconds((updateInterval * 2));
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.GONE,
```
ThingStatusDetail.OFFLINE.GONE is probably not appropriate. Rather simply COMMUNICATION_ERROR ?

lolodomo reviewed (changes requested):
Strange things with the thing config/proerties and with background discovery

splatch commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/discovery/SMAEnergyMeterDiscoveryService.java:
```diff
@@ -54,35 +62,52 @@ public Set<ThingTypeUID> getSupportedThingTypes() {
-        properties.put(Thing.PROPERTY_SERIAL_NUMBER, energyMeter.getSerialNumber());
-        ThingUID uid = new ThingUID(THING_TYPE_ENERGY_METER, energyMeter.getSerialNumber());
+        properties.put(Thing.PROPERTY_SERIAL_NUMBER, identifier);
+        ThingUID uid = new ThingUID(THING_TYPE_ENERGY_METER, identifier);
```
Identifier is a number turned into String so additional formatting is not needed. If device brings invalid characters for identifier I'd rather let framework mark thing as invalid instead of  accepting it.

splatch commented:
@lolodomo I've fixed troubles you pointed out, other issues you reported I leave for further discussion.

splatch commented:
Note - this PR is a breaking change as it introduces a new parameter to thing (serialNumber).

Homer-Sim commented:
Hello, 
is the fix for this binding now OK and available?

splatch commented:
There were some notes on it in related forum thread. Beyond these code worked fine (when I had access to SMA equipment) and filtered out wrong packet kinds.

lsiepel commented:
@splatch this PR has both awaiting feedback and WIP label, are they still valid or should a maintainer look at it?

splatch commented:
I've updated PR to latest main, please re-review it as it solves two major troubles for users and one which affects quality of data delivered by energy meter. This PR been open for almost two years during which I lost access to installation where I could test binding.

FYI #13437 from @lrep proposed some good improvements for protocol handling which would clarify a bit binding codebase.

lsiepel commented:
Can you also run the i18n plugin. 

If i understand this PR, the breaking change is only the mandatory configuration (no channels, or functional changes) so i think a update notice would be enough. Please create a PR here: https://github.com/openhab/openhab-distro/pulls

splatch commented:
> @splatch this PR has both awaiting feedback and WIP label, are they still valid or should a maintainer look at it?

I haven't added any of these labels, thus I am not a valid person to answer this question.

lsiepel commented:
@splatch as you didn;t add the WIP label and stated before this is ready for review, i removed the WIP label. But before a merge can be made, there are 3-4 comments from lolodomo that are not yet marked as resolved. Could you look at them (they are deeply hidden behind all the load more buttons)

Could you also add a breaking change notice here: https://github.com/openhab/openhab-distro/pulls

splatch commented:
@lsiepel I've updated copyright headers, i18n definitions and removed part of discovery service which was in question in unresolved comments from 2022. It should be good to go once CI finishes its job.

kaikreuzer commented:
@cambronf Not sure why you are pinging me here. As far as I can read from above, the last updates were done due to @lolodomo's review comments, so he should check if they are addressed as wished. Then it is for any @openhab/add-ons-maintainers to merge the PR.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/DefaultPacketListenerRegistry.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    public ScheduledFuture<?> addTask(Runnable runnable, int intervalSec) {
+        return scheduler.scheduleAtFixedRate(runnable, 0, intervalSec, TimeUnit.SECONDS);
```
https://www.openhab.org/docs/developer/guidelines.html#e-runtime-behavior

scheduleWithFixedDelay is preferred to scheduleAtFixedRate. is scheduleAtFixedRate absolutely required here ?

lsiepel replied:
I guess `scheduleWithFixedDelay` is good enough, maybe @splatch can comment if the far more expensive `scheduleAtFixedRate` is essential.

lsiepel replied:
Changed.

splatch replied:
Sorry, I am stuck in preparations to industrial trade fair which takes place next week. Meter side push data every 100-200 ms (AFAIR), so ideally we should really pick them and publish at fixed rate. The fixed delay proposed in official docs is wrong in many places, cause it introduces random delays which are out-of-control for anyone. 

In context of this particular change it is irrelevant, cause it is not for receiving data, but publishing which is effortless and takes little to almost no-time.

lolodomo replied:
PR was merged with `scheduleWithFixedDelay` before I saw your message.

If this is really important, a new PR can be submitted.

splatch replied:
No problem, its fine to remain as-is.

lsiepel replied:
Thanks for assuring this is ok!

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PayloadHandler.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @author Łukasz Dywicki - Initial contribution
+ */
+public interface PayloadHandler {
```
Please add `@NonNullByDefault`

lsiepel replied:
done

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/DefaultPacketListenerRegistry.java:
```diff
@@ -0,0 +1,85 @@
+ * @author Łukasz Dywicki - Initial contribution
+ */
+@Component
+public class DefaultPacketListenerRegistry implements PacketListenerRegistry {
```
Please add `@NonNullByDefault`

lsiepel replied:
done

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java:
```diff
@@ -0,0 +1,130 @@
+ *
+ * @author Łukasz Dywicki - Initial contribution
+ */
+public class PacketListener {
```
Please add `@NonNullByDefault`

lsiepel replied:
done

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java:
```diff
@@ -0,0 +1,130 @@
+public class PacketListener {
+
+    private final DefaultPacketListenerRegistry registry;
+    private final List<PayloadHandler> handlers = new CopyOnWriteArrayList<>();
```
It is better to use a `Set`, it is a precaution to avoid having the same object several times in the list.
You can find examples in other bindings. Let me know if you need help to find one.

lsiepel replied:
Please advice. i don't see any example.

lolodomo replied:
It was more an advice, not a request for change. It can work well with a List if the binding is properly implemented.
I remember we got memory leaks in certain bindings due to that. I have in mind I fixed that in elerotransmitterstick binding few years ago but I can't find the PR...
As it is just a suggestion, you can ignore it.

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java:
```diff
@@ -0,0 +1,130 @@
+    void close() throws IOException {
+        if (future != null) {
+            future.cancel(true);
+        }
```
You can also set `future` to null.

lsiepel replied:
done

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java:
```diff
@@ -0,0 +1,130 @@
+        InetAddress address = InetAddress.getByName(multicastGroup);
+        socket.leaveGroup(address);
+        socket.close();
+    }
```
You can also set `socket` to null.

lsiepel replied:
done

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java:
```diff
@@ -0,0 +1,130 @@
+                        handler.handle(meter);
+                    }
+                } catch (IOException e) {
+                    logger.info("Unexpected payload received for group {}", group, e);
```
Rather DEBUG level.

lsiepel replied:
done

lolodomo reviewed (changes requested):
Full review done

lolodomo commented:
I am asking myself why there is the tag "potentially not backward compatible" ? My feeling is that your code is backward compatible. Of course, after the change, all existing things will be OFFLINE until the serial number is filled.

lsiepel commented:
Gentle ping @splatch can you continue ?

splatch commented:
@lsiepel  I've reflected my view on this here: https://community.openhab.org/t/sma-energy-meter-binding-yields-unplausible-values/128180/165

To phrase it again, I'm not going to waste more time on nit picks. I'm can fix the socket leak, but nothing else.

lsiepel commented:
> not going to waste more

Ah, got a bit late to the party. Sorry to hear this somewhat negative pov. I would not call them nit picks, but i can understand that it sometimes feels like it. A project like this has some guidelines and rules not to irritate you, but to assure quality, maintainability and so on. I  also understand that a long standing PR like this is another factor in the equasion.
Anyway, i would really appreciate it if you take the effort to fix the review comments. If not i can try to commit to your repository to get this accross the finish line. Please let me know.

splatch commented:
@lsiepel I am not irritated, because I do not use this binding, hence I do not suffer from missed functionality.  The most recent review contributed nothing (or at least very little) to quality of this addon, these are rather personal opinions of reviewer. More over these contradict purpose of Java GC and openHAB framework itself. If user controllable log output in one place or setting null values after dispose, empty method body, or way in which default values are set, are not nit picks, then what you consider a such?

Code I've submitted for review passed quality checks enforced by SAT. Neither of matters from recent review represent issue for SAT, if it should, then please update SAT to save not only reviewer but also other contributors time. 
Its not the first nor the latest PR, stuck by mysterious quality criteria combined with _I do not like this construction™_ review. 

There is issue with socket leak reported for ipv6 already addressed in forums, which is hardly related to recent comments.

splatch commented:
@lsiepel None of review checklist mention empty method bodies, sets over lists. Use of primitives over complex types is subjective matter related to nullnes, which in above case when elements are mapped by config mapper, is not wrong. I probably could find dozen of such use cases in existing codebase, yet nobody fight with them.

I'll fix the socket leak if I can. I won't touch nit picks. You decide whether you do for users, continuous improvement or continuous waiting.

lsiepel commented:
> @lsiepel None of review checklist mention empty method bodies, sets over lists. Use of primitives over complex types is subjective matter related to nullnes, which in above case when elements are mapped by config mapper, is not wrong. I probably could find dozen of such use cases in existing codebase, yet nobody fight with them.
> 
> I'll fix the socket leak if I can. I won't touch nit picks. You decide whether you do for users, continuous improvement or continuous waiting.

Tried to fix all comments by a uploading a commit to your repository but permissions are lacking. Could you allow me to perform a commit?

lsiepel commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/discovery/SMAEnergyMeterDiscoveryService.java:
```diff
@@ -54,35 +62,43 @@ public Set<ThingTypeUID> getSupportedThingTypes() {
-            logger.debug("Diagnostic: ", e);
-            return;
-        }
+    @Override
+    public void startScan() {
+    }
```
I think it is mandatory to have this as `SMAEnergyMeterDiscoveryService extends AbstractDiscoveryService`
When it is removed i get a build error. If still needs to be removed, please advice.

lolodomo replied:
Yes, correct, it is an abstract method.

lsiepel commented:
I now have triage access, but no write access, could you check again, thanks.

splatch commented:
@lsiepel done, you should have write access now.

lsiepel commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +68,81 @@ public void initialize() {
-        pollingJob = scheduler.scheduleWithFixedDelay(this::updateData, 0, pollingPeriod, TimeUnit.SECONDS);
-        logger.debug("Polling job scheduled to run every {} sec. for '{}'", pollingPeriod, getThing().getUID());
-
-        updateStatus(ThingStatus.ONLINE);
```
I guess this is fine now as it is set by the listener, similar to how a scheduler would do.

lsiepel commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +68,81 @@ public void initialize() {
-            pollingJob.cancel(true);
-            pollingJob = null;
+        if (listener != null) {
+            listener.removePayloadHandler(this);
```
done

lsiepel commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/DefaultPacketListenerRegistry.java:
```diff
@@ -0,0 +1,85 @@
+
+    private final Logger logger = LoggerFactory.getLogger(DefaultPacketListenerRegistry.class);
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1,
+            (runnable) -> new Thread(runnable, "sma-energy-meter-receiver"));
```
Done

lsiepel commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +68,81 @@ public void initialize() {
 
-        int port = (config.getPort() == null) ? EnergyMeter.DEFAULT_MCAST_PORT : config.getPort();
-        energyMeter = new EnergyMeter(config.getMcastGroup(), port);
+        int port = (config.getPort() == null) ? PacketListener.DEFAULT_MCAST_PORT : config.getPort();
```
done

lsiepel commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +68,81 @@ public void initialize() {
+
         try {
-            energyMeter.update();
+            listener = listenerRegistry.getListener(config.getMcastGroup(), port);
```
done

lsiepel commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -61,68 +68,81 @@ public void initialize() {
+    }
+
+    @Nullable
+    private static String retrieveSerialNumber(Map<String, String> properties) {
```
done

lsiepel commented:
If i understood it right, @splatch has some details about a leak and might fix this. I fixed most review comments and have some questions remaining @lolodomo Hopefully we can get this to the finish line. Thanks!

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/discovery/SMAEnergyMeterDiscoveryService.java:
```diff
@@ -54,35 +65,48 @@ public Set<ThingTypeUID> getSupportedThingTypes() {
+    protected void stopBackgroundDiscovery() {
+        PacketListener packetListener = this.packetListener;
+        if (packetListener != null) {
+            packetListener.removePayloadHandler(this);
```
I guess it would be clean to also set this.packetListener  to null ?

lsiepel replied:
done

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java:
```diff
@@ -35,94 +37,103 @@
 
     @Override
     public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
```
Put this line rather at line 84, this will avoid a doubled thing state change in case one parameter is missing, UNKNOWN and finally OFFLINE few ms after.

lsiepel replied:
done

lolodomo commented on bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java:
```diff
@@ -0,0 +1,145 @@
+
+    static class ReceivingTask implements Runnable {
+        private final Logger logger = LoggerFactory.getLogger(ReceivingTask.class);
+        private final @Nullable DatagramSocket socket;
```
Maybe it is better to have it there not nullable and to check if the socket parameter is not null when instantiating the class ?

lsiepel replied:
Don't think this matters much, but done.

lolodomo reviewed (changes requested):
Final comments

lolodomo commented:
Only remains the question about usage of `scheduleAtFixedRate`.

lolodomo commented:
No answer?

lsiepel commented:
> No answer?

Seems not, i changed it to `scheduleWithFixedDelay` as i'm pretty sure that a little variation in the receiving of packets is not going to harm anything. Please continue with review/merging. So we can put all this behind us.

lolodomo reviewed (approved):
LGTM

lolodomo commented:
> i changed it to `scheduleWithFixedDelay` as i'm pretty sure that a little variation in the receiving of packets is not going to harm anything. Please continue with review/merging. So we can put all this behind us.

Thank you @lsiepel for having considered all comments

cambronf commented:
The smaenergymeter binding released in 4.2.0M3 shows a huge build up of UDP multicast frames, causing delays of up to half an hour between the real measurment and the reported value in openhab.
This issue has been reported earlier this year:
[https://community.openhab.org/t/sma-energy-meter-binding-yields-unplausible-values/128180/132?u=fcambron]

The broadcast frames fix introduces clearly a new issue, keeping the bindung un-usable.

```
openhabian@openhabian:~ $ netstat -nau
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
udp6  165888      0 :::9522                 :::*
udp6  165888      0 :::9522                 :::*
udp6  156672      0 :::9522                 :::*
udp6  172032      0 :::9522                 :::*
udp6  165888      0 :::9522                 :::*
udp6  159744      0 :::9522                 :::*
udp6  153600      0 :::9522                 :::*
udp6  181248      0 :::9522                 :::*
```

lolodomo commented:
This will be clearly better to create a new issue. If not, your message could be simply not seen or forgotten in a short time.
